include/ensmallen_bits/function/add_evaluate.hpp

```diff
@@ -60,6 +60,42 @@ class AddEvaluate<FunctionType,
     return static_cast<FunctionType*>(
         static_cast<Function<FunctionType,
                              MatType, GradType>*>(this))->Evaluate(coordinates);
+   }
+
+  typename MatType::elem_type Evaluate(arma::mat predictors, arma::mat responses)
+  {
+    return static_cast<FunctionType*>(
+        static_cast<Function<FunctionType,
+                             MatType, GradType>*>(this))->Evaluate(predictors, responses);
+  }
+};
+
+/**
+ * Reflect the existing Evaluate().
+ */
+template<typename FunctionType,
+         typename MatType,
+         typename GradType>
+class AddEvaluate<FunctionType,
+                  MatType,
+                  GradType,
+                  false,
+                  false>
+{
+ public:
+  // Reflect the existing Evaluate().
+  typename MatType::elem_type Evaluate(const MatType& coordinates)
+  {
+    return static_cast<FunctionType*>(
+        static_cast<Function<FunctionType,
+                             MatType, GradType>*>(this))->Evaluate(coordinates);
```

Comment:
I think this is one of the reasons why @rcurtin proposed to implement this in mlpack. I'm wondering if we can put this into the callback itself, instead of extending the Function Type interface, which we like to keep to a minimum.

Reply:
@zoq I agree. However, I tried to implement it first in mlpack, but in mlpack case, we will have to change the architecture of this callback.
I think it is the same, even if we are in mlpack we will to `add_evaluate` in ensmallen. If we had a close look, it seems to me that `EndEpoch()` requires `FunctionType& function` in our case,  which is `ens::Function`. Suppose `ens::Function</*mlpack neural network  model should goes here*/>::Evaluate(arma::mat, arma::mat)` interface does not exist. Can you give me more details about possible implementation in mlpack?

Otherwise, I can try to move the overloaded interface in the callbacks.

Reply:
@rcurtin Do you have any thoughts on this, any starting guide from mlpack?

Reply:
I think that maybe the best way to put this together would be to implement a callback that has structure a little bit like this:

```
class EarlyStoppingOnValidationSet // maybe can be named better?
{
 private:
  FFN<>& network;
  arma::mat& validationPredictors;
  arma::mat& validationResponses;
  double lastResponse; // initialize to DBL_MAX?

 public:
  template<typename OptimizerType, typename FunctionType, typename MatType>
  bool EndEpoch(OptimizerType& /* optimizer */,
                FunctionType& /* function */,
                const MatType& /* coordinates */,
                const size_t /* epoch */,
                const double /* objective */)
  {
    // The FFN we are holding internally should be the same one as what is being optimized.
    const double validationLoss = ffn->Evaluate(validationPredictors, validationResponses);
    if (validationLoss > lastLoss)
       return true; // or do we return false to terminate early?  I can't remember

    return false;
  }
};
```

I've omitted some functions there, but maybe that's a decent enough idea?  It would definitely have to be in the mlpack codebase since it depends on `FFN`.  Let me know what you think---and sorry it took so long to get to this. :+1:

Reply:
Now, I understand the idea related to the `FFN` reference. I will implement it and see if it works.
Thanks. :+1: